File: 215/c/main.c
#include <limits.h>

typedef struct {
    int size;
    int count;
    int* array;
} Heap;

void siftUp(Heap* heap, int index) {
    while (index > 0) {
        int parent_index = (index-1)/2;
        int parent = heap->array[parent_index];
        int child = heap->array[index];
        if (parent < child) break;

        heap->array[index] = parent;
        heap->array[parent_index] = child;
        index = parent_index;
    }
}

void siftDown(Heap* heap, int index) {
    int size = heap->size;
    int count = heap->count;
    for (int child_index = index * 2 + 1; child_index < size; child_index = index * 2 + 1) {
        int child = heap->array[child_index];
        if (count > child_index + 1 && heap->array[child_index + 1] < child) {
            child = heap->array[++child_index];
        }

        int parent = heap->array[index];
        if (parent <= child) break;

        heap->array[index] = child;
        heap->array[child_index] = parent;
        index = child_index;
    }
}

void insert(Heap* heap, int num) {
    int end = heap->count;
    if (end < heap->size) {
        heap->array[end] = num;
        siftUp(heap, end);
        heap->count++;
    } else if (heap->array[0] < num) {
        heap->array[0] = num;
        siftDown(heap, 0);
    }
}

int findKthLargest(int* nums, int numsSize, int k) {

    if (numsSize == 1) { 
        return nums[0];
    }

    if (k == 1) {
        int largest = INT_MIN;
        for (int i = 0; i < numsSize; i++) largest = nums[i] > largest ? nums[i] : largest;
        return largest;
    }

    int array[k];
    Heap seen = { .size = k, .count = 0, .array = array };
    for (int i = 0; i < numsSize; i++) insert(&seen, nums[i]);

    return array[0];
}

File: 215/python/main.py
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        seen = Heap(k)
        for num in nums:
            seen.insert(num)
        return seen.top()

class Heap:
    def __init__(self, size: int):
        self.size = size
        self.count = 0
        self.array = [None] * size

    def top(self) -> int:
        return self.array[0]

    def insert(self, num: int):
        if self.count < self.size:
            self.array[self.count] = num
            self.sift_up(self.count)
            self.count += 1
        else:
            if self.array[0] < num:
                self.array[0] = num
                self.sift_down(0)

    def sift_up(self, index: int):
        while index > 0:
            swap_index = (index-1)//2
            parent = self.array[swap_index]
            child = self.array[index]
            if parent < child:
                break

            self.array[index] = parent
            self.array[swap_index] = child
            index = swap_index
            
    def sift_down(self, index: int):
        while index*2+1 < self.size:
            child_index = (index*2)+1
            child = self.array[child_index]
            if self.count > (child_index + 1) and self.array[child_index+1] < child:
                child_index += 1
                child = self.array[child_index]

            parent = self.array[index]
            if parent > child:
                self.array[index] = child
                self.array[child_index] = parent
                index = child_index
            else:
                break
